File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Effect/Combat/ShockWaveEffect.ts
import { FairyGUI } from "csharp";
import { Mathf } from "../../../../../Core/Module/Math/Mathf";
import Color from "../../DataDefine/Color";
import DungeonManager from "../../DungeonManager";
import { AbstractEffect } from "../AbstractEffect";
import { ShakeScreen, ShakeIntensity, ShakeDur } from "../ShakeScreen";
import { BlurWaveChaoticEffect } from "./BlurWaveChaoticEffect";
import { BlurWaveEffect } from "./BlurWaveEffect";


export class ShockWaveEffect extends AbstractEffect{
    public constructor(private _X:number,private _Y:number,private _Color:Color,private _ShockWaveType:ShockWaveType) {
        super();
    }

    Update() {
        super.Update();
        let speed = Mathf.RandomRange(1000,1200);
        switch(this._ShockWaveType){
            case ShockWaveType.ADDITIVE: {
                for(let i = 0;i<40;i++){
                    DungeonManager.EffectManager.Play(new BlurWaveEffect(this._X,this._Y,this._Color,speed,FairyGUI.BlendMode.Add));
                }
                break;
            }
            case ShockWaveType.NORMAL:{
                for(let i = 0;i<40;i++){
                    DungeonManager.EffectManager.Play(new BlurWaveEffect(this._X,this._Y,this._Color,speed,FairyGUI.BlendMode.Normal));
                }
                break;
            }
            case ShockWaveType.CHAOTIC:{
                DungeonManager.EffectManager.Play(new ShakeScreen(ShakeIntensity.HIGH,ShakeDur.SHORT,{Vertical: false,Horizontal: true}))
                for(let i = 0;i<40;i++){
                    DungeonManager.EffectManager.Play(new BlurWaveChaoticEffect(this._X,this._Y,this._Color,speed));
                }
            }
        }
    }
}

export enum ShockWaveType {
    ADDITIVE,
    NORMAL,
    CHAOTIC
}

File: Unity/Assets/Scripts/ModTheSpire/Effect/Combat/ShockWaveEffect.cs
using Core.Utils.Unity;
using ModTheSpire.Dungeon;
using UnityEngine;

namespace ModTheSpire.Effect.Combat
{
    public class ShockWaveEffect : AbstractEffect
    {
        private float mX;
        private float mY;
        private Color mColor;
        private ShockWaveType mShockWaveType;

        public ShockWaveEffect(float x, float y, Color color, ShockWaveType shockWaveType)
        {
            mX = x;
            mY = y;
            mColor = color;
            mShockWaveType = shockWaveType;
        }

        public override void Update()
        {
            base.Update();
            var speed = Random.Range(1000,1201);
            switch(this.mShockWaveType){
                case ShockWaveType.ADDITIVE: {
                    for(var i = 0;i<40;i++){
                        DungeonManager.EffectManager.Play(new BlurWaveEffect(this.mX,this.mY,this.mColor,speed,FairyGUI.BlendMode.Add));
                    }
                    break;
                }
                case ShockWaveType.NORMAL:{
                    for(var i = 0;i<40;i++){
                        DungeonManager.EffectManager.Play(new BlurWaveEffect(this.mX,this.mY,this.mColor,speed,FairyGUI.BlendMode.Normal));
                    }
                    break;
                }
                case ShockWaveType.CHAOTIC:
                {
                    CameraUtils.Shake(ShakeVibrato.HIGH, ShakeDur.SHORT, new Vector3(0.1f, 0));
                    for(var i = 0;i<40;i++){
                        DungeonManager.EffectManager.Play(new BlurWaveChaoticEffect(this.mX,this.mY,this.mColor,speed));
                    }

                    break;
                }
            }
        }
    }
    
    public enum ShockWaveType {
        ADDITIVE,
        NORMAL,
        CHAOTIC
    }
}